flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.ColumnFamilyOptions;
+import org.rocksdb.CompactionStyle;
+import org.rocksdb.DBOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+import org.rocksdb.StringAppendOperator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link InternalTimerService} that stores timers in RocksDB.
+ */
+public class RocksDBInternalTimerService<K, N> extends InternalTimerService<K, N> {
+	
+	private static Logger LOG = LoggerFactory.getLogger(RocksDBInternalTimerService.class);
+	
+	/** The data base where stores all timers */
+	private final RocksDB db;
+	
+	/** The path where the rocksdb locates */
+	private final Path dbPath;
+
+	/**
+	 * The in-memory heaps backed by rocksdb to retrieve the next timer to trigger. Each
+	 * partition's leader is stored in the heap. When the timers in a partition is changed, we
+	 * will change the partition's leader and update the heap accordingly.
+	 */
+	private final int numPartitions;
+	private final PersistentTimerHeap eventTimeHeap;
+	private final PersistentTimerHeap processingTimeHeap;
+	
+	private static int MAX_PARTITIONS = (1 << 16);
+
+	public RocksDBInternalTimerService(
+			int totalKeyGroups,
+			KeyGroupRange keyGroupRange,
+			KeyContext keyContext,
+			ProcessingTimeService processingTimeService,
+			Path dbPath) {
+
+		super(totalKeyGroups, keyGroupRange, keyContext, processingTimeService);
+		
+		this.dbPath = dbPath;
+		
+		try {
+			FileSystem fileSystem = this.dbPath.getFileSystem();
+			if (fileSystem.exists(this.dbPath)) {
+				fileSystem.delete(this.dbPath, true);
+			}
+			
+			fileSystem.mkdirs(dbPath);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while creating directory for rocksdb timer service.", e);
+		}
+
+		ColumnFamilyOptions columnFamilyOptions = new ColumnFamilyOptions()
+				.setMergeOperator(new StringAppendOperator())
+				.setCompactionStyle(CompactionStyle.UNIVERSAL);
+		ColumnFamilyDescriptor defaultColumnDescriptor = new ColumnFamilyDescriptor("default".getBytes(), columnFamilyOptions);
+
+		DBOptions dbOptions = new DBOptions()
+				.setCreateIfMissing(true)
+				.setUseFsync(false)
+				.setDisableDataSync(true)
+				.setMaxOpenFiles(-1);
+
+		List<ColumnFamilyHandle> columnFamilyHandles = new ArrayList<>(1);
+
+		try {
+			this.db = RocksDB.open(dbOptions, dbPath.getPath(), Collections.singletonList(defaultColumnDescriptor), columnFamilyHandles);
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while creating the RocksDB instance.", e);
+		}
+
+		this.numPartitions = Math.min(keyGroupRange.getNumberOfKeyGroups(), MAX_PARTITIONS);
```

Comment:
I think we can safely initialize this to `keyGroupRange.getNumberOfKeyGroups()`, which is required to be smaller than the maximum max_parallelism. A preconditions check might be ok, though.